### DISCUSSION TITLE
Iterate file name by one if it already exists

### DISCUSSION THREAD

**Discussion by Assblastois** [SilentVoid13/Templater#1185] [Help] [answered]
Hi there :) 

I need some help. I am trying to make a template which renames the file automatically in this format: “Meeting YYYY-MM-DD No. 1”. This works great when I have one meeting, but I run into trouble when I have two or more on the same day. I would use quickadd but it doesn’t work in my case because I create the meetings note in a projects folder to generate backlinks to that project. I then automatically move it to a dedicated Meetings folder. 
This is the code I have come up with, but it does not work. 
My idea was to check the folder if the file name already existed. If so, iterate the number by one, check again, etc. until the file name does not exist, and break the loop. 

This currently does not work. It iterates the “let meetingNumber” by only one, and then breaks even if the name does already exist. 
I would really appreciate some help. If there is more info you need, please let me know. 

Here is the code: 
```
const meetingsFolder = 'Meetings/';
const today = tp.date.now('YYYY-MM-DD');

let meetingNumber = 1
let meetingName = `Meeting ${today} No. ${meetingNumber}`;


while (tp.file.exists(meetingsFolder + meetingName)) {
	meetingNumber++
	meetingName = `Meeting ${today} No. ${meetingNumber}`;
		if (! await tp.file.exists(meetingsFolder + meetingName)){
		console.log(meetingName)
			break;
		}else{
		console.log("no")
			continue;
		}
}

await tp.file.rename(meetingName)
```

**Comment by Zachatoo** [comment, marked as answer]
I played with your script a bit to try to get it working with the while loop, but I couldn't get it to work. The problem is is that `tp.file.exists()` is async, and I guess they don't work well for the condition in a while loop.

I did find an alternate solution. Instead of looping, I'm grabbing the files in the meeting folder, filtering down to only meetings that happened today, sorting, then incrementing the number by the latest meeting number.

```js
<%*
const meetingsFolder = 'Meetings';
const today = tp.date.now('YYYY-MM-DD');
let meetingNumber = 1;

// Get list of all meetings today
const todaysMeetings = app
  .vault
  .getAbstractFileByPath(meetingsFolder)
  .children
  .filter(child => child instanceof tp.obsidian.TFile && child.basename.startsWith(`Meeting ${today}`));
// If there are any meetings that already happened today,
// set meeting number to last meeting number plus 1
if (todaysMeetings.length) {
  todaysMeetings.sort((a, b) => a.path < b.path ? 1 : -1);
  const latestMeeting = todaysMeetings[0];
  meetingNumber = parseInt(latestMeeting.basename.match(/\d+$/)[0], 10) + 1;
}

const meetingName = `Meeting ${today} No. ${meetingNumber}`;
await tp.file.rename(meetingName)
-%>
```

**Comment by Assblastois** [comment]
Wow... this works like a charm. I was thinking of maybe doing a sort, but I have no programing knowledge and even writing my shitty script took me hours. Thank you so so much. 
This is my quickadd setup for this template. I add the move command to the end of the note with the capture feature. 

I had run into problems of quickadd refusing to make changes to the file after moving it, giving me an error because I tried to modify the file again. (Something like that.)
I guess I can also use your code to make something similar with my daily note generator. 
I can’t thank you enough.

**Reply by Assblastois** [reply]
![image](https://github.com/SilentVoid13/Templater/assets/143261073/d15e6057-cddc-4c54-a2fe-ad4e73699c05)

**Comment by Assblastois** [comment]
@Zachatoo Hi again 😅 

So, I am still using your code for my meetings, daily notes (for today and tomorrow via tp.date.now and .tomorrow) and it workes great. I now thought: Why not just use it with Natural Language Dates' Date Picker? 
This sadly does not work and it pastes the date inside the note. The title is not correct either. It also causes problems with quickadd. Is it because the code can't utilize the outputted 'anyday' value? 

(i can't resize the damn image)

![image](https://github.com/SilentVoid13/Templater/assets/143261073/1c85a3ac-1030-4cfb-b9d6-282f3d5c25a5)


```<%*
const DailyNotesFolder = 'Daily notes';
const anyday =  app.commands.executeCommandById('nldates-obsidian:nlp-picker');;
let dailyNoteNumber = 1;

// Get list of all dates today
const anydaysDailyNote = app
  .vault
  .getAbstractFileByPath(DailyNotesFolder)
  .children
  .filter(child => child instanceof tp.obsidian.TFile && child.basename.startsWith(`${anyday}`));
// If there are any dates that already happened today,
// set date number to last date number plus 1
if (anydaysDailyNote.length) {
  anydaysDailyNote.sort((a, b) => a.path < b.path ? 1 : -1);
  const latestDailyNote = anydaysDailyNote[0];
  dailyNoteNumber = parseInt(latestDailyNote.basename.match(/\d+$/)[0], 10) + 1;
}

const dailyNoteName = `${anyday} No. ${dailyNoteNumber}`;
await tp.file.rename(dailyNoteName)
-%>```

**Reply by Zachatoo** [reply]
We'll have to dig into the minified source code to see why this doesn't work. Here's the minified code for `app.commands.executeCommandById` and `app.commands.executeCommand`.

```js
e.prototype.executeCommandById = function(e, t) {
    var n = this.findCommand(e);
    return !!n && this.executeCommand(n, t)
}
,
e.prototype.executeCommand = function(e, t) {
    this.app.lastEvent = t || null;
    try {
        o_(e)
    } catch (t) {
        return console.log("Command failed to execute: ", e.id),
        console.error(t),
        !1
    }
    return !0
}
```

`!1` is equivalent to `false`, and `!0` is equivalent to `true`. So `app.commands.executeCommand` will always return `true` if the command succeeds, and `false` if it fails. Same with `app.commands.executeCommandById`. **You can't get the return value from the command** if there was one. That is why your code doesn't work, `anyday` will always return either `true` or `false`.

```js
const anyday =  app.commands.executeCommandById('nldates-obsidian:nlp-picker');
```

**Reply by Assblastois** [reply]
That's a shame, but I already thought it might be impossible. Thank you for looking into it. It would have been a nice to have.
The script calls the date picker function outside the script, so there is no way it could store the resulting value if I understand it correctly.

**Reply by Zachatoo** [reply]
You'd have to make a PR into that other plugin to send an event that Templater can pickup and do something with. In it's current state, it would not be possible.
